torch/nn/utils/spectral_norm.py

```diff
@@ -53,8 +53,14 @@ def apply(module, name, n_power_iterations, eps):
 
         u = normalize(weight.new_empty(height).normal_(0, 1), dim=0, eps=fn.eps)
         delattr(module, fn.name)
-        module.register_parameter(fn.name + "_org", weight)
-        module.register_buffer(fn.name, weight)
+        module.register_parameter(fn.name + "_orig", weight)
+        # We still need to assign weight back as fn.name because all sorts of
+        # things may assume that it exists, e.g., when initializing weights.
+        # However, we can't directly assign as it could be an nn.Parameter and
+        # gets added as a parameter. Instead, we assign weight.data, which will
+        # just be added as plain attribute, and also supports nn.init due to
+        # shared storage.
+        setattr(module, fn.name, weight.data)
```

Comment:
I must admit admit that I'm not sure that calling init_ on a replaced parameter is something to support - for example it would work before the first forward but not afterwards...
That said, other options aren't necessarily better (of course, when it's late and I'm tired I wonder whether the separation of weight and weight_orig is really that important).

Reply:
Yeah, it won't work after the first forward. :( 
Separation of weight and weight_orig is unfortunately necessary with the current interface. In the hook we will have to replace weight with the new normalized value, but we still need to keep the old weight around to optimize it (and we can't assign it as .weight).